micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java

```diff
@@ -56,32 +82,123 @@ public MongoMetricsCommandListener(MeterRegistry registry) {
      * @param tagsProvider provides tags to be associated with metrics for the given Mongo command
      * @since 1.7.0
+     * @deprecated use {@link MongoMetricsCommandListener.Builder} instead
      */
+    @Deprecated
     public MongoMetricsCommandListener(MeterRegistry registry, MongoCommandTagsProvider tagsProvider) {
-        this.registry = registry;
-        this.tagsProvider = tagsProvider;
+        this(MongoMetricsCommandListener.builder(registry).tagsProvider(tagsProvider));
+    }
+
+    private MongoMetricsCommandListener(Builder builder) {
+        this.registry = builder.registry;
+        this.timerMetricName = builder.timerMetricName;
+        this.timerMetricDesc = builder.timerMetricDescription;
+        this.tagsProvider = builder.tagsProvider;
+        this.includeCollectionNameTag = builder.includeCollectionNameTag;
+        if (builder.collectionNamesCacheMaxSize < 1) {
+            throw new IllegalArgumentException("collectionNamesCacheMaxSize must be set to a positive value");
+        }
+        this.collectionNamesCacheMaxSize = builder.collectionNamesCacheMaxSize;
+        this.collectionNamesCacheOverflowErrorTracker = builder.collectionNamesCacheOverflowErrorLogInterval < 1 ? null : new EveryNthOperation(builder.collectionNamesCacheOverflowErrorLogInterval);
     }
 
     @Override
-    public void commandStarted(CommandStartedEvent commandStartedEvent) {
+    public void commandStarted(CommandStartedEvent event) {
+        if (!includeCollectionNameTag) {
+            return;
+        }
+        mongoCommandUtil.determineCollectionName(event.getCommandName(), event.getCommand())
+                .ifPresent(collectionName -> addCollectionNameToCache(event, collectionName));
     }
 
     @Override
     public void commandSucceeded(CommandSucceededEvent event) {
-        timeCommand(event, event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
     @Override
     public void commandFailed(CommandFailedEvent event) {
-        timeCommand(event, event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
-    private void timeCommand(CommandEvent event, long elapsedTimeInNanoseconds) {
-        Timer.builder("mongodb.driver.commands")
-                .description("Timer of mongodb commands")
-                .tags(tagsProvider.commandTags(event))
-                .register(registry)
+    private void timeCommand(CommandEvent event, String collection, long elapsedTimeInNanoseconds) {
+        Timer.Builder timerBuilder = Timer.builder(this.timerMetricName)
+                .description(this.timerMetricDesc)
+                .tags(this.tagsProvider.commandTags(event));
+        if (this.includeCollectionNameTag) {
+            timerBuilder.tag("collection", collection);
```

Comment:
Can we move most of the new code introduced in this file into the `tagsProvider`?

Reply:
Ahh yeh. Now I remember. This code (the collectionName tag) was originally included in my contribution for the tags provider MR we did. Because it complicated the code, I wanted to push the simpler pieces through first and then follow on in a subsequent MR (this current one). I will go back and refactor this to fit into the TagsProvider model we added. 👍🏻 


```suggestion
            timerBuilder.tag("collection", collection);
```